File: sdk/core/unity/Runtime/Core/NetworkMonitor.cs
using System;
using System.Collections.Generic;
using UnityEngine;
using UnityEngine.Networking;

namespace RivalApex.SDK.Core
{
    /// <summary>
    /// NetworkMonitor - Monitors network connectivity and provides fast-fail behavior.
    ///
    /// This class provides:
    /// - Real-time network state monitoring using Unity's Network APIs
    /// - Fast-fail for no-network conditions (no UI jank)
    /// - Connection quality assessment
    /// - Callback-based network state change notifications
    /// </summary>
    public sealed class NetworkMonitor : MonoBehaviour
    {
        private static NetworkMonitor _instance;
        private static readonly object _lock = new object();
        
        /// <summary>
        /// Singleton instance of NetworkMonitor
        /// </summary>
        public static NetworkMonitor Instance
        {
            get
            {
                if (_instance == null)
                {
                    lock (_lock)
                    {
                        if (_instance == null)
                        {
                            var go = new GameObject("RivalApexNetworkMonitor");
                            _instance = go.AddComponent<NetworkMonitor>();
                            DontDestroyOnLoad(go);
                        }
                    }
                }
                return _instance;
            }
        }
        
        /// <summary>
        /// Fast-fail timeout when offline (seconds)
        /// </summary>
        public const float OfflineFastFailTimeout = 0.1f;
        
        /// <summary>
        /// Normal network timeout (seconds)
        /// </summary>
        public const float NormalTimeout = 10f;
        
        /// <summary>
        /// Interval between connectivity checks (seconds)
        /// </summary>
        public float CheckInterval = 3f;
        
        /// <summary>
        /// Connection type enumeration
        /// </summary>
        public enum ConnectionType
        {
            None,
            Wifi,
            Cellular,
            Ethernet,
            Other
        }
        
        /// <summary>
        /// Network state information
        /// </summary>
        public readonly struct NetworkState : IEquatable<NetworkState>
        {
            public readonly bool IsConnected;
            public readonly ConnectionType ConnectionType;
            public readonly bool IsExpensive;
            public readonly DateTime Timestamp;
            
            public NetworkState(bool isConnected, ConnectionType connectionType, bool isExpensive)
            {
                IsConnected = isConnected;
                ConnectionType = connectionType;
                IsExpensive = isExpensive;
                Timestamp = DateTime.UtcNow;
            }
            
            public static NetworkState Offline => new NetworkState(false, ConnectionType.None, false);
            
            public bool Equals(NetworkState other)
            {
                return IsConnected == other.IsConnected && 
                       ConnectionType == other.ConnectionType && 
                       IsExpensive == other.IsExpensive;
            }
            
            public override bool Equals(object obj) => obj is NetworkState other && Equals(other);
            
            public override int GetHashCode() => HashCode.Combine(IsConnected, ConnectionType, IsExpensive);
            
            public static bool operator ==(NetworkState left, NetworkState right) => left.Equals(right);
            public static bool operator !=(NetworkState left, NetworkState right) => !left.Equals(right);
        }
        
        /// <summary>
        /// Result of a pre-flight check before network operations
        /// </summary>
        public readonly struct PreflightResult
        {
            public readonly bool ShouldProceed;
            public readonly string FailReason;
            public readonly NetworkState State;
            
            private PreflightResult(bool shouldProceed, string failReason, NetworkState state)
            {
                ShouldProceed = shouldProceed;
                FailReason = failReason;
                State = state;
            }
            
            public static PreflightResult Proceed(NetworkState state) => 
                new PreflightResult(true, null, state);
            
            public static PreflightResult FastFail(string reason, NetworkState state) => 
                new PreflightResult(false, reason, state);
        }
        
        /// <summary>
        /// Delegate for network state change events
        /// </summary>
        public delegate void NetworkStateChangedHandler(NetworkState state);
        
        /// <summary>
        /// Event fired when network state changes
        /// </summary>
        public event NetworkStateChangedHandler OnNetworkStateChanged;
        
        private NetworkState _currentState = NetworkState.Offline;
        private readonly Dictionary<string, NetworkStateChangedHandler> _listeners = new();
        private float _lastCheckTime;
        private bool _isMonitoring;
        
        /// <summary>
        /// Gets the current network state
        /// </summary>
        public NetworkState State => _currentState;
        
        /// <summary>
        /// Checks if network is currently connected
        /// </summary>
        public bool IsConnected => _currentState.IsConnected;
        
        /// <summary>
        /// Checks if on an expensive connection
        /// </summary>
        public bool IsExpensive => _currentState.IsExpensive;
        
        private void Awake()
        {
            if (_instance != null && _instance != this)
            {
                Destroy(gameObject);
                return;
            }
            
            _instance = this;
            DontDestroyOnLoad(gameObject);
            UpdateNetworkState();
        }
        
        /// <summary>
        /// Starts monitoring network state changes
        /// </summary>
        public void StartMonitoring()
        {
            _isMonitoring = true;
            UpdateNetworkState();
        }
        
        /// <summary>
        /// Stops monitoring network state changes
        /// </summary>
        public void StopMonitoring()
        {
            _isMonitoring = false;
        }
        
        private void Update()
        {
            if (!_isMonitoring) return;
            
            if (Time.unscaledTime - _lastCheckTime >= CheckInterval)
            {
                _lastCheckTime = Time.unscaledTime;
                UpdateNetworkState();
            }
        }
        
        /// <summary>
        /// Performs a preflight check before network operations.
        /// Returns immediately if offline, allowing fast-fail behavior.
        /// </summary>
        public PreflightResult Preflight()
        {
            // Force update before preflight
            UpdateNetworkState();
            
            if (_currentState.IsConnected)
            {
                return PreflightResult.Proceed(_currentState);
            }
            
            return PreflightResult.FastFail("No network connection", _currentState);
        }
        
        /// <summary>
        /// Gets the appropriate timeout based on network state.
        /// Returns fast-fail timeout when offline.
        /// </summary>
        public float EffectiveTimeout()
        {
            return IsConnected ? NormalTimeout : OfflineFastFailTimeout;
        }
        
        /// <summary>
        /// Adds a listener for network state changes.
        /// </summary>
        /// <returns>A token to use when removing the listener</returns>
        public string AddListener(NetworkStateChangedHandler listener)
        {
            var token = Guid.NewGuid().ToString();
            _listeners[token] = listener;
            
            // Immediately notify with current state
            listener?.Invoke(_currentState);
            return token;
        }
        
        /// <summary>
        /// Removes a listener using its token.
        /// </summary>
        public void RemoveListener(string token)
        {
            _listeners.Remove(token);
        }
        
        /// <summary>
        /// Gets network quality hints for adaptive behavior.
        /// </summary>
        public Dictionary<string, object> QualityHints()
        {
            return new Dictionary<string, object>
            {
                ["isConnected"] = _currentState.IsConnected,
                ["connectionType"] = _currentState.ConnectionType.ToString(),
                ["isExpensive"] = _currentState.IsExpensive,
                ["suggestedTimeout"] = EffectiveTimeout(),
                ["shouldReduceQuality"] = _currentState.IsExpensive
            };
        }
        
        /// <summary>
        /// Forces an immediate network state update
        /// </summary>
        public void ForceUpdate()
        {
            UpdateNetworkState();
        }
        
        private void UpdateNetworkState()
        {
            var reachability = Application.internetReachability;
            
            ConnectionType connectionType;
            bool isConnected;
            bool isExpensive;
            
            switch (reachability)
            {
                case NetworkReachability.ReachableViaLocalAreaNetwork:
                    connectionType = ConnectionType.Wifi;
                    isConnected = true;
                    isExpensive = false;
                    break;
                    
                case NetworkReachability.ReachableViaCarrierDataNetwork:
                    connectionType = ConnectionType.Cellular;
                    isConnected = true;
                    isExpensive = true;
                    break;
                    
                case NetworkReachability.NotReachable:
                default:
                    connectionType = ConnectionType.None;
                    isConnected = false;
                    isExpensive = false;
                    break;
            }
            
            var newState = new NetworkState(isConnected, connectionType, isExpensive);
            
            if (newState != _currentState)
            {
                _currentState = newState;
                NotifyListeners();
            }
        }
        
        private void NotifyListeners()
        {
            OnNetworkStateChanged?.Invoke(_currentState);
            
            foreach (var listener in _listeners.Values)
            {
                listener?.Invoke(_currentState);
            }
        }
        
        private void OnDestroy()
        {
            if (_instance == this)
            {
                _instance = null;
            }
        }
    }
    
    /// <summary>
    /// Exception for fast-fail on no-network conditions
    /// </summary>
    public class NoNetworkException : Exception
    {
        public NetworkMonitor.NetworkState NetworkState { get; }
        
        public NoNetworkException(string message, NetworkMonitor.NetworkState networkState)
            : base(message)
        {
            NetworkState = networkState;
        }
    }
}

File: sdk/core/ios/Sources/Network/NetworkMonitor.swift
import Foundation
import Network

/// NetworkMonitor - Monitors network connectivity and provides fast-fail behavior.
///
/// This class provides:
/// - Real-time network state monitoring using NWPathMonitor
/// - Fast-fail for no-network conditions (no UI jank)
/// - Connection quality assessment
/// - Callback-based network state change notifications
@available(iOS 12.0, tvOS 12.0, macOS 10.14, *)
public final class NetworkMonitor {
    
    /// Shared instance for convenience
    public static let shared = NetworkMonitor()
    
    /// Fast-fail timeout when offline (seconds)
    public static let offlineFastFailTimeout: TimeInterval = 0.1
    
    /// Normal network timeout (seconds)
    public static let normalTimeout: TimeInterval = 10.0
    
    /// Network state information
    public struct NetworkState: Equatable {
        public let isConnected: Bool
        public let connectionType: ConnectionType
        public let isExpensive: Bool
        public let isConstrained: Bool
        public let timestamp: Date
        
        public static let offline = NetworkState(
            isConnected: false,
            connectionType: .none,
            isExpensive: false,
            isConstrained: false,
            timestamp: Date()
        )
    }
    
    /// Connection type enumeration
    public enum ConnectionType: String {
        case none
        case wifi
        case cellular
        case ethernet
        case loopback
        case other
    }
    
    /// Result of a pre-flight check before network operations
    public enum PreflightResult {
        /// Network is available, proceed with operation
        case proceed(NetworkState)
        
        /// Network unavailable, fast-fail with this error
        case fastFail(reason: String, state: NetworkState)
    }
    
    /// Listener closure type for network state changes
    public typealias NetworkStateListener = (NetworkState) -> Void
    
    private let monitor: NWPathMonitor
    private let queue = DispatchQueue(label: "com.rivalapex.networkmonitor", qos: .utility)
    
    private var currentState = NetworkState.offline {
        didSet {
            if oldValue != currentState {
                notifyListeners()
            }
        }
    }
    
    private var listeners: [UUID: NetworkStateListener] = [:]
    private let listenersLock = NSLock()
    
    private var isMonitoring = false
    
    public init() {
        self.monitor = NWPathMonitor()
        setupMonitor()
    }
    
    private func setupMonitor() {
        monitor.pathUpdateHandler = { [weak self] path in
            self?.updateState(from: path)
        }
    }
    
    /// Starts monitoring network state changes
    public func startMonitoring() {
        guard !isMonitoring else { return }
        isMonitoring = true
        monitor.start(queue: queue)
    }
    
    /// Stops monitoring network state changes
    public func stopMonitoring() {
        guard isMonitoring else { return }
        isMonitoring = false
        monitor.cancel()
    }
    
    /// Gets the current network state
    public var state: NetworkState {
        return currentState
    }
    
    /// Checks if network is currently connected
    public var isConnected: Bool {
        return currentState.isConnected
    }
    
    /// Checks if we're on an expensive connection (cellular data)
    public var isExpensive: Bool {
        return currentState.isExpensive
    }
    
    /// Performs a preflight check before network operations
    /// Returns immediately if offline, allowing fast-fail behavior
    public func preflight() -> PreflightResult {
        let state = currentState
        
        if state.isConnected {
            return .proceed(state)
        } else {
            return .fastFail(reason: "No network connection", state: state)
        }
    }
    
    /// Gets the appropriate timeout based on network state
    /// Returns fast-fail timeout when offline
    public func effectiveTimeout() -> TimeInterval {
        return isConnected ? Self.normalTimeout : Self.offlineFastFailTimeout
    }
    
    /// Adds a listener for network state changes
    /// - Returns: A token to use when removing the listener
    @discardableResult
    public func addListener(_ listener: @escaping NetworkStateListener) -> UUID {
        let token = UUID()
        listenersLock.lock()
        listeners[token] = listener
        listenersLock.unlock()
        
        // Immediately notify with current state
        listener(currentState)
        return token
    }
    
    /// Removes a listener using its token
    public func removeListener(token: UUID) {
        listenersLock.lock()
        listeners.removeValue(forKey: token)
        listenersLock.unlock()
    }
    
    /// Gets network quality hints for adaptive behavior
    public func qualityHints() -> [String: Any] {
        let state = currentState
        return [
            "isConnected": state.isConnected,
            "connectionType": state.connectionType.rawValue,
            "isExpensive": state.isExpensive,
            "isConstrained": state.isConstrained,
            "suggestedTimeout": effectiveTimeout(),
            "shouldReduceQuality": state.isExpensive || state.isConstrained
        ]
    }
    
    private func updateState(from path: NWPath) {
        let connectionType: ConnectionType
        if path.usesInterfaceType(.wifi) {
            connectionType = .wifi
        } else if path.usesInterfaceType(.cellular) {
            connectionType = .cellular
        } else if path.usesInterfaceType(.wiredEthernet) {
            connectionType = .ethernet
        } else if path.usesInterfaceType(.loopback) {
            connectionType = .loopback
        } else if path.status == .satisfied {
            connectionType = .other
        } else {
            connectionType = .none
        }
        
        currentState = NetworkState(
            isConnected: path.status == .satisfied,
            connectionType: connectionType,
            isExpensive: path.isExpensive,
            isConstrained: path.isConstrained,
            timestamp: Date()
        )
    }
    
    private func notifyListeners() {
        let state = currentState
        listenersLock.lock()
        let currentListeners = listeners
        listenersLock.unlock()
        
        for (_, listener) in currentListeners {
            DispatchQueue.main.async {
                listener(state)
            }
        }
    }
}

/// Error type for fast-fail on no-network conditions
public struct NoNetworkError: Error {
    public let message: String
    public let networkState: NetworkMonitor.NetworkState
    
    public init(message: String, networkState: NetworkMonitor.NetworkState) {
        self.message = message
        self.networkState = networkState
    }
}

/// Fallback for older iOS versions without NWPathMonitor
public final class LegacyNetworkMonitor {
    
    public static let shared = LegacyNetworkMonitor()
    
    /// Simple reachability check via URLSession
    public func checkConnectivity(timeout: TimeInterval = 2.0, completion: @escaping (Bool) -> Void) {
        guard let url = URL(string: "https://www.apple.com/library/test/success.html") else {
            completion(false)
            return
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "HEAD"
        request.timeoutInterval = timeout
        
        let task = URLSession.shared.dataTask(with: request) { _, response, _ in
            if let httpResponse = response as? HTTPURLResponse {
                completion(httpResponse.statusCode == 200)
            } else {
                completion(false)
            }
        }
        task.resume()
    }
    
    /// Synchronous connectivity check (use with caution - blocks thread)
    public func isConnectedSync(timeout: TimeInterval = 0.5) -> Bool {
        let semaphore = DispatchSemaphore(value: 0)
        var connected = false
        
        checkConnectivity(timeout: timeout) { result in
            connected = result
            semaphore.signal()
        }
        
        _ = semaphore.wait(timeout: .now() + timeout + 0.1)
        return connected
    }
}
